src/MPMCQueue.NET/MPMCQueue.cs

```diff
@@ -4,88 +4,113 @@
+    [StructLayout(LayoutKind.Explicit, Size = 336)]
     public class MPMCQueue
     {
-        [FieldOffset(0)]
-        private readonly Cell[] _buffer;
-        [FieldOffset(8)]
-        private readonly int _bufferMask;
-        [FieldOffset(64)]
-        private int _enqueuePos;
-        [FieldOffset(128)]
-        private int _dequeuePos;
+        /// <summary>
+        /// 128 bytes cache line already exists in some CPUs.
+        /// </summary>
+        /// <remarks>
+        /// Also "the spatial prefetcher strives to keep pairs of cache lines in the L2 cache."
+        /// https://stackoverflow.com/questions/29199779/false-sharing-and-128-byte-alignment-padding
+        /// </remarks>
+        internal const int SAFE_CACHE_LINE = 128;
 
+        [FieldOffset(SAFE_CACHE_LINE)]
+        private readonly Cell[] _enqueueBuffer;
+
+        [FieldOffset(SAFE_CACHE_LINE + 8)]
+        private volatile int _enqueuePos;
+
+        // Separate access to buffers from enqueue and dequeue.
+        // This removes false sharing and accessing a buffer 
+        // reference also prefetches the following Pos with [(64 - (8 + 4 + 4)) = 52]/64 probability.
+
+        [FieldOffset(SAFE_CACHE_LINE * 2)]
+        private readonly Cell[] _dequeueBuffer;
+
+        [FieldOffset(SAFE_CACHE_LINE * 2 + 8)]
+        private volatile int _dequeuePos;
 
         public MPMCQueue(int bufferSize)
         {
             if (bufferSize < 2) throw new ArgumentException($"{nameof(bufferSize)} should be greater than or equal to 2");
             if ((bufferSize & (bufferSize - 1)) != 0) throw new ArgumentException($"{nameof(bufferSize)} should be a power of 2");
 
-            _bufferMask = bufferSize - 1;
-            _buffer = new Cell[bufferSize];
+            _enqueueBuffer = new Cell[bufferSize];
 
             for (var i = 0; i < bufferSize; i++)
             {
-                _buffer[i] = new Cell(i, null);
+                _enqueueBuffer[i] = new Cell(i, null);
             }
 
+            _dequeueBuffer = _enqueueBuffer;
             _enqueuePos = 0;
             _dequeuePos = 0;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public bool TryEnqueue(object item)
         {
+            var spinner = new SpinWait();
             do
             {
-                var buffer = _buffer;
+                var buffer = _enqueueBuffer;
                 var pos = _enqueuePos;
-                var index = pos & _bufferMask;
-                var cell = buffer[index];
+                var index = pos & (buffer.Length - 1);
+                ref var cell = ref buffer[index];
                 if (cell.Sequence == pos && Interlocked.CompareExchange(ref _enqueuePos, pos + 1, pos) == pos)
                 {
-                    buffer[index].Element = item;
-                    Volatile.Write(ref buffer[index].Sequence, pos + 1);
+                    cell.Element = item;
+                    cell.Sequence = pos + 1;
```

Comment:
`Volatile.Write` is [now almost identical to direct](https://github.com/dotnet/runtime/issues/4363) reads/writes to volatile fields.

On line 62 `cell.Sequence` is now a volatile read which is fairly slow on ARM. 
I don't know[ if RuyJIT has changed implementation to use ARMv8 half-barries](https://github.com/dotnet/runtime/issues/6314), but they are not availabe onARMv7  anyway.

Reply:
I am still only planning to buy Raspberry PI for testing :) Yes, good point. It's next to Interlocked call anyways. I changed that mostly from aesthetics perspective.